Feature/ghela alert blogpost (#628)
ordina-jworks/ordina-jworks.github.io: _posts/2020-09-30-ghelamco-alert.md

---
layout: post
authors: [kevin_govaerts, bas_moorkens]
title: 'Ghelamco alert: '

> image: /img/2020-09-25-ghelamco-alert/rpi-front.jpg

Reviewer: this does not have the correct size: 500x293

If the parking lot isn't cleared in time we risk fines up to 500 Euros per car.  
Of course we don't want to spend our money on fines when instead we could be buying more pizzas.  
So we came up with a solution called **Ghelamco alert**.  
The idea for the project is pretty straightforward.  
We will run a Raspberry PI device that is hooked up to an alert light.  
The Raspberry PI will run an application that parses the game fixtues on the website of KAA Gent to search for any home fixtures.  
Whenever a home fixture occurs, the RPI will start turning on the alert light on a preset schedule.  
This way the employees in the office will have a visual warning on game day and they will be reminded to leave the parking lot on time.  
Since we wanted to build a user friendly solution we added a serverless backend in the AWS cloud combined with an angular application so that our users can also look at the web application to get some additional feedback and even manipulate the alerting schedule.  
This enables users to do some additional operations through the web application:
* Snooze alerts
* Create additional alerts
* Create custom events

> <!-- //TODO GIF Proof of working project -->

Reviewer: TODO?

As the last step we just need to add the H2 dependency to our projects pom.xml .  
```
<dependency>
    <groupId>com.h2database</groupId>
	<artifactId>h2</artifactId>
	<scope>runtime</scope>
</dependency>
```

Add some configuration in the application.properties file.  
```
### H2 Storage
spring.datasource.url=jdbc:h2:file:c:\\h2-ghela\\h2db
spring.datasource.driverClassName=org.h2.Driver
spring.datasource.username=kevin
spring.datasource.password=admin
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
spring.jpa.hibernate.ddl-auto=create-drop
spring.h2.console.enabled=true
spring.h2.console.settings.web-allow-others=false
```

This just works out of the box, it even comes with a console to check your database if enabled.

<div style="text-align: center;">
  <img alt="Ghelamco-alert Cloudwatch Alarm" src="/img/2020-09-25-ghelamco-alert/h2-console.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

#### Metrics
{:.no_toc}

Our application generates metrics to provide us with some data about our backend.  

> We collect data and push it into AWS cloudwatch as metrics for:

Reviewer: Add a link to the aws cloudWatch page

* Number of events added

> * Number of events update

Reviewer: ```suggestion
* Number of events updated
```

* Number of events cancelled
* Number of alerts snoozed
* Number of alerts started
* Number of alerts stopped
* A basic heartbeat

> When our application is running, we are constantly gathering this data which we push to Cloudwatch metrics, every 5 minutes.

Reviewer: ```suggestion
When our application is running, we are constantly gathering this data which we push to Cloudwatch metrics every 5 minutes.
```

> Afterwards we reset our data counters, and start collecting again.  

Reviewer: ```suggestion
Afterwards we reset our data counters and start collecting again.  
```

This allows us to create a dashboard which has data-points every 5 minutes for all the metrics we defined.  

### Connecting our RPI to the cloud
We have two ways of connecting back to the AWS cloud.  
* Use AWS credentials and the AWS apis.  
* Use AWS IoT MQTT protocol and the according certificates.

#### AWS Cloudwatch integration
{:.no_toc}

The cloudwatch integration is done entirely by having a set of credentials with limited permissions on our RPI device.  
We use these credentials inside our application to access the AWS SDK from our java code.  
We integrate with cloudwatch in two different ways:
* We push custom metrics from our application to cloudwatch metrics.
* We send the logs from our application to cloudwatch logs.

##### Cloudwatch metrics
{:.no_toc}

The metrics that we generate in our application should of course get sent to AWS cloudwatch, so we can actually make use of them in the AWS cloud.  

> We used the metric data to build dashboards and alerts within cloudwatch for added visibility into our application.  

Reviewer: ```suggestion
We use the metric data to build dashboards and alerts within cloudwatch for added visibility into our application.  
```

</div>

We have created an alarm that triggers when our **RPI_BACKEND_STATUS** status metric is missing for 3 data points in a row.  
This means that after 15 minutes of not receiving this metric, the cloudwatch alarm will trigger and notifies us via email that the RPI is either down or disconnected from the internet.  
This enables us to respond quickly and take action to restore connectivity to the device.  

<div style="text-align: center;">
  <img alt="Ghelamco-alert Cloudwatch Alarm" src="/img/2020-09-25-ghelamco-alert/ghelamco-rpi-backend-down-alert.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

##### Cloudwatch logs
{:.no_toc}

The logs from our spring boot application get sent to Cloudwatch logs. 
This allows us to check the log files in the AWS cloud without needing access to the RPI device itself.  
<div style="text-align: center;">
  <img alt="Ghelamco-alert Cloudwatch Alarm" src="/img/2020-09-25-ghelamco-alert/cloudwatch-logs.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

#### AWS IoT
{:.no_toc}

> The reason for using AWS IoT is for the sake of "being connected to the cloud".  

Reviewer: add a link to aws iot

When we are connected to the cloud, we are able to communicate with our device "through the cloud", meaning from anywhere we want!   
In our case this means that we want our RPI device to be connected to AWS IoT whenever it is up and running and has internet connectivity.  
Of course the communication between our RPI and the AWS cloud has to be secured.
AWS IoT uses authentication and authorization workflow by using certificates that you issue from AWS IoT and upload to your edge device.  
We treat our RPI as an edge device in this project.

Authentication works based on the certificates that the edge device presents to AWS IoT.  
Once the certificate authentication is successful, AWS IoT checks which policies got attached to those certificates to grant it specific authorizations within the AWS cloud.  

> ##### Authentication: certificates

Reviewer: lots of things in this section

{:.no_toc}

When you add a new edge device in AWS IOT it is called "a thing".  
When registering "a thing" it generates 2 keys for us: a public key and a private key.  
AWS IoT will provide a certificate signing request for the public key, which will sign the generated certificate with the root certificate's private key.  

Our thing-certificate and our private key are our credentials when we try to communicate with AWS IoT to access our edge device.  
The only additional input that we need to provide on top of the generated signing request is the root certificate to check the signing.  

* Certificates
* private Key
* Root CA

Since these certificates get generated per thing that you register they should only be used for 1 device.  
When I developed the application I had to register another thing for my laptop.  

> Connecting from the RPI and the laptop with the same certificates caused some unwanted behavior like connection interrupts.  

Reviewer: Can this be detected in aws iot?

Would you be able to take over a connection by having another device send in data?

Just some food for thought

Author: yeah, if you posses the certificates you have the keys to the connection. best practice is probably not to share certificates but it was easier this way in our dev phase

<div style="text-align: center;">
  <img alt="Ghelamco-alert Cloudwatch Alarm" src="/img/2020-09-25-ghelamco-alert/cert-aws-iot.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

##### Authorization: policies
{:.no_toc}

> After creating our certificates and keys, we need to handle the authorization part.  

Reviewer: ```suggestion
After creating our certificates and keys we need to handle the authorization part.  
```

We do this by adding some policies to our certificates.  
When presenting our certificates on connect, AWS IoT now also knows what services we can access within the AWS cloud. 

An example of such a policy: 
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "iot:Publish",
        "iot:Subscribe",
        "iot:Connect",
        "iot:Receive"
      ],
      "Resource": [
        "*"
      ]
    },
    {
      "Effect": "Allow",
      "Action": [
        "iot:GetThingShadow",
        "iot:UpdateThingShadow",
        "iot:DeleteThingShadow"
      ],
      "Resource": [
        "*"
      ]
    },
    {
      "Effect": "Allow",
      "Action": [
        "greengrass:*"
      ],
      "Resource": [
        "*"
      ]
    }
  ]
}
```

We define which services we want to access:
- "iot:Publish"
- "iot:Subscribe
- "iot:GetThingShadow"
- "iot:UpdateThingShadow"
- "iot:DeleteThingShadow" 

This system of generating certificates and coupling policies is a very secure and easy way of working with edge devices.   

##### jobs
{:.no_toc}

A crucial part of AWS IoT is the job section. 

> We can create a jobs by using the AWS cli.   

Reviewer: a job or jobs?

```
aws iot update-job  
  --job-id 010  
  --description "updated description" 
  --timeout-config inProgressTimeoutInMinutes=100
  --job-executions-rollout-config {exponentialRate:{baseRatePerMinute: 50, incrementFactor: 2, rateIncreaseCriteria: <{numberOfNotifiedThings: 1000, numberOfSucceededThings: 1000}, maximumPerMinute: 1000}}
  --abort-config { criteriaList: [ { action: CANCEL, failureType: FAILED, minNumberOfExecutedThings: 100, thresholdPercentage: 20}, { action: CANCEL, failureType: TIMED_OUT, minNumberOfExecutedThings: 200, thresholdPercentage: 50}]}          
  --presigned-url-config {roleArn:arn:aws:iam::123456789012:role/S3DownloadRole, expiresInSec:3600}
```
We use these jobs AWS Iot to create these jobs and send them to our RPI where they get executed by our application code.  

<div style="text-align: center;">
  <img alt="Ghelamco-alert device update" src="/img/2020-09-25-ghelamco-alert/device-update-iot.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

We have the possibility to put multiple "things" in a group of devices.  

> This way we can send jobs to the entire group and the job gets executed on every device from that group.  

Reviewer: ```suggestion
This way we can send jobs to the entire group and these jobs get executed on every device from that group.  
```

These jobs could be software updates, reboot commands, rotation of certificates, ...   
Anything we want really!  

In our project we used IoT jobs for a multitude of processes:  
* Update our backend application - over the air
* Create new events
* Snooze and update existing alerts

<div style="text-align: center;">
  <img alt="Ghelamco-alert Cloudwatch Alarm" src="/img/2020-09-25-ghelamco-alert/aws-iot-jobs.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

It is pretty straightforward to create an IoT job:

1. Create a job
2. Add a job document (JSON file) which defines the content of the job 
3. Push the job onto the job queue to send it to your devices

Example job document:  
```
{
    "operation": "snoozeAlert",
    "local_alert_id": 31,
    "event_title": "test from lambda snooze alert"
}
```

##### MQTT protocol
{:.no_toc}

> AWS IoT can communicate with it's registered devices through 2 protocols: HTTP or MQTT.  

Reviewer: ```suggestion
AWS IoT can communicate with its registered devices through 2 protocols: HTTP or MQTT.  
```

* HTTP is a 1-1 protocol. The client makes a request and the server responds.  
It is difficult and expensive to broadcast a message to all devices on the network, which is a common use case in IoT applications.
* HTTP is a heavyweight protocol with many headers and rules. It is not suitable for constrained networks.

MQTT on the other hand defines two types of entities in the network: a message broker (AWS IoT) and a number of edge devices(clients).  
The broker is a server that receives all messages from the clients and then routes those messages to relevant destination clients.  
A client is anything that can interact with the broker to send and receive messages.  
Our client is the RPI but it could also be an IoT sensor in the field or an application in a data center that processes IoT data.  

In our backend java code we have an MQTTJobService which connects to AWS IoT by using the [AWS SDK](https://aws.amazon.com/sdk-for-java/) and subscribes to the relevant topics to receive and respond to AWS Iot jobs.  
Every 30 seconds we will read these topics to see if there are any new jobs to be processed.  

### AWS IOT Greengrass

> AWS greengrass is a service that extends the AWS cloud onto your edge device.  

Reviewer: add a link to greengrass

This was fascinating for us since we had some tough problems to solve:  
* How can we deploy our application on the RPI device?  
* How can we make sure our application recovers from failures?

> * How can we keep our system itself up to date?

Reviewer: ```suggestion
* How can we make sure that our system keeps itself up to date?
```

AWS greengrass offers solutions to all these challenges!  
Greengrass is an extension of the AWS IoT service and since we had already set up our device in AWS IoT it was easy for us to set up greengrass.  
To get started with greengrass we had to do 2 additional steps:
* Define a greengrass group. This group will contain your IoT devices and deployments.  
* Define a greengrass core. The core is the device that you will use to run the additional AWS capabilities on.  

The greengrass group allows us to push deployments to all the devices in our group.  
It also allows you to run lambda functions on your core device or install additional connectors with your own runtime of choice.  
<div style="text-align: center;">
  <img alt="Greengrass group" src="/img/2020-09-25-ghelamco-alert/gg_group.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

When you register your device as a core device in greengrass you immediately get some nice additional benefits from this.  
For example, you can immediately see all the network interfaces on your core device and what ip addresses got allocated.  

> This is especially useful if you want to ssh to your device and do not have fixed ip attached to it.  

Reviewer: ```suggestion
This is especially useful if you want to ssh to your device and do not have a fixed ip attached to it.  
```

<div style="text-align: center;">
  <img alt="Greengrass core" src="/img/2020-09-25-ghelamco-alert/gg_core.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

> Now we that greengrass is installed and our RPI is configured as a core device our greengrass group we can start making full use of the capabilities that greengrass offers.  

Reviewer: ```suggestion
Now that greengrass is installed and our RPI is configured as a core device on our greengrass group, we can start making full use of the capabilities that greengrass offers.  
```

> We decided that we wanted our application to run as a docker container, so we installed and configured the **docker connector** for greengrass.   

Reviewer: Add a link to that docker connector page

like: https://docs.aws.amazon.com/greengrass/latest/developerguide/docker-app-connector.html

This plugin allows you to run docker containers on your core device and makes use of **docker** and **docker compose**.  
<div style="text-align: center;">
  <img alt="Greengrass docker connector" src="/img/2020-09-25-ghelamco-alert/gg_connector_docker_cfg.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>
For this to work you have to install docker and docker compose on your core device.  
Since we are using a RPI it was a bit harder to install since the RPI uses the ARM7 chipset instead of x64/x32.  
We had no trouble installing docker itself, but for docker compose we had to clone the source code from github and compile the binaries ourself on the RPI.  
After all the setup was done we could just create a docker container in our CICD pipeline and tell greengrass to run that container on RPI.  

### CICD pipeline RPI Backend

To make sure we did not need to bother ourselves with manual builds and installs of our code on the RPI we built a CICD pipeline to automatically deploy our software onto the RPI.  

> We trigger the pipeline whenever a push to our master branch in our git repository.  

Reviewer: ```suggestion
We trigger the pipeline whenever a commit is pushed to our master branch in our git repository.  
```

> We used azure devops as our CICD system.  

Reviewer: link to azure devops

Note: why azure devops and doesn't aws have something similar?

Author: because bousson has an azure fetish

<div style="text-align: center;">
  <img alt="Ghelamco-alert Azure pipeline backend" src="/img/2020-09-25-ghelamco-alert/azure-backpipe.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

The pipeline has a lot of pipeline steps defined and integrates with several AWS services as well.  
We drew up a chart to provide additional insight into how this pipeline works:
<div style="text-align: center;">
  <img alt="Ghelamco-alert Azure pipeline backend" src="/img/2020-09-25-ghelamco-alert/ghela-alert-cicd-rpi.png" width="auto" height="auto" target="_blank" class="image fit">
</div>

As you can see there are 2 major parts of this pipeline: 
* the azure devops pipeline
* the greengrass deployment process.  

#### Azure devops pipeline
{:.no_toc}

* Build and test our java code.  
We use maven to build and test our code.   

> * We use [AWS SSM - parameter store]() to safely store our config files and certificates for the RPI.  

Reviewer: missing link to aws ssm?

In step 2 and 3 we fetch those config files and secrets from the parameter store and store them locally for use later in the pipeline.  

> * In order to control the alert light hooked up to our RPI we need to download the wiringPi library as we used that to communicate with our alert light. 

Reviewer: add a link to the wiringPi library

> We stored this library in our releases S3 bucket.  

Reviewer: releases or release?

Otherwise, azure devops would have no way of running our docker image since it cannot interpret the ARM7 processor instruction set natively.  

In our **docker-compose.yml** file we describe how the container should behave when it's running:
```
version: '3.3'
services:
  web:
    image: "${ecr.imageTag}"
    privileged: true
    ports:
      - "8080:8080"
    volumes:
      - /home/pi/.aws/:/root/.aws:ro
      - /ghela/db:/ghela/db
volumes:
  db-data:
```
Several things are happening in this file:
* We dynamically inject the image name and version in our cicd pipeline into the  **${ecr.imageTag}** field.  
* We run our container in privileged mode as this is needed to access the RPIs native interface to control our alert light through GPIO with wiringPi.
* We expose our application port 8080 to the outside world.
* We mount our AWS credentials and H2 database as volumes from the RPI host system.

#### Greengrass deployment process
{:.no_toc}

For the greengrass deployment process we did not have to create any code.  
This works out of the box when you set up greengrass.  
However, for full transparency I will briefly describe the process here:

> * AWS greengrass receives a new deployment ( this is the last step of our azure devops pipeline).

Reviewer: ```suggestion
* AWS greengrass receives a new deployment (this is the last step of our azure devops pipeline).
```

All these steps get done automatically for us by the greengrass agent running on our RPI device.  
As you can see this CICD pipeline makes it incredibly easy for us to deploy new software to our RPI device.  

## Frontend web application

To get data from our RPI into our web application we needed a way to connect to our H2 database on the RPI.  
It would be pretty complex to set up our RPI device to be accessible from the internet, so we chose to build a backend in AWS to function as a proxy for our RPI backend application.  
This allows us to use this proxy backend in AWS to access the data from our RPI device and send new commands to update existing data on the RPI.  
More about this backend in the serverless part.  
First let's take a look into our frontend application.  

### Angular
I wanted to create a simple web application to get more insight in the data of our RPI device.  
Any modern frontend framework would be suitable for this but I decided to go with Angular since I had already had some basic experience with it.  
We did however use some extra frameworks and AWS services to make it easier to build our frontend application.  

### AWS Amplify

> To help us bootstrap our web application, we decided to use the AWS amplify framework.  

Reviewer: add a link to aws amplify

In the job overview view we query our api gateway to fetch us all the pending jobs from our dynamoDB job table.  
The lambda function that listens to this endpoint reads the ghela-jobs dynamoDB table and list all jobs that have a status of not yet "COMPLETED".  
The moment this list is empty all changes from our webapp have been processed successfully by our RPI backend application and will have updated the ghela-events and ghela-jobs dynamoDB tables.  

Important note: our RPI can only receive new jobs from AWS IoT if it is connected to the internet.  
So if we do not have a working network connection jobs will stay queued within AWS IoT and not get pushed to the RPI backend application.  
For the same reason it is important to know that our RPI device is connected and functioning well, to make sure our commands from our web application get processed by our RPI backend application.  

<div style="text-align: center;">
  <img alt="Ghelamco-alert jobs screen" src="/img/2020-09-25-ghelamco-alert/frontend-jobs.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div>

#### Metrics
{:.no_toc}

On this tab we can see the status of the RPI backend as a graph.  

> Remember the Metrics we talked about before?  

Reviewer: ```suggestion
Remember the metrics we talked about before?  
```

Bas then introduced me to the concept of [IaC or Infrastructure as Code](https://en.wikipedia.org/wiki/Infrastructure_as_code) and [the Serverless framework](https://www.serverless.com/).  
There are several options to do IaC on AWS but for our particular project serverless seemed the best fit.  
Our entire cloud backend gets configured in our **serverless.yml** file, that is versioned on github in our cloud backend project.  
This way we can set up our infrastructure through IaC in a repeatable manner.  

### Serverless.yml

This file is the heart of our serverless setup.  
Everything that is needed to run our serverless stack got described in this file.  
We used the **yaml** notation as it is less convoluted than JSON, but you can also use JSON or typescript to build this.   
We will explain the different parts of our serverless setup to show you in detail how everything works. 

#### Functions and Api Gateway
{:.no_toc}

The essential part of our serverless setup is the functions section.  
A function is an AWS Lambda function.  
It's an independent unit of deployment like a true microservice.  
It is generally a very small piece of code that does one thing and does it well.  

> In our project for example we have a lambda function **list-events** which does exactly that, list the events from our dynamoDB table.  

Reviewer: ```suggestion
In our project, for example, we have a lambda function **list-events** which does exactly that, list the events from our dynamoDB table.  
```

1. We click snooze alert in the frontend web application
2. This calls our api gateway endpoint **"/alert/snooze"**.
3. This call is authorized by the **userpool JWT token** and **identitypool temporary AWS credential**.  
4. This api gateway endpoint invokes an **AWS Lambda function** which creates a job and **job document in AWS iot** for the RPI.
It also inserts a row into a second **dynamoDB table** (ghela-jobs) to register the job we created on the frontend.  
5. The job gets send to the RPI and processed by our backend application, thus snoozing the alert in the **H2 database**.
This will effectively turn the alert light off.    
6. The RPI **synchronizes** his H2 database with the event **dynamoDB table**. Thus updating the status of the alert in the dynamoDB table.  
7. This new status will then be reflected in the **frontend web application** so our user can see the result of his action.  

This same workflow applies to all commands that can be sent from the frontend application to our RPI device.  
This is a prime example of how a distributed system works and operates.  

# Conclusion

> This internship gave me such a big cover of all best practices and new technologies, that this was a real eye-opening and educational experience.  

Reviewer: ```suggestion
This internship gave me such a big overview of all best practices and new technologies, that this was a real eye-opening and educational experience.  
```

Or: This internship covered such a large amount of best practices and new technologies that this was a real eye-opening educational experience.

- I've had the chance to develop an application in Spring, with different databases.
- learned how to properly test my code with unit and integration tests
- learned how to use the Java and Node AWS SDKs
- API's, connecting it all together...
- came in touch with a lot of linux, sharpening my commandline skills
- developed an Angular app from start to finish
- learned a lot about AWS: IAM, IoT Core, DynamoDB, API Gateway, SNS, SSM, EKS, S3, Cloudwatch, Cloudfront, lambdas, Route53, ...
- CICD pipelines and their best practices in the industry

I'm very thankful for being able to do this and having the chance to work with Bas on a real project.  

He always motivated me to go the extra mile !  

> He could destroy a day of hard work in a few sentences... :)  

Reviewer: dat is hoe dat we hem kennen 😉

Author: haha :)

> He could destroy a day of hard work in a few sentences... :)  

Reviewer: ```suggestion
And he could destroy a day of hard work in a few sentences ... :)  
```

<div style="text-align: center;">

>   <img alt="Ghelamco-alert Cloudwatch Alarm" src="/img/2020-09-25-ghelamco-alert/bas-smile.png" width="auto" height="auto" target="_blank" >

Reviewer: image not perfectly cut out, maybe put it on a white background to remove the pixelated effect at the borders?

</div>

> But afterwards, he would always take the time to explain why, and how I should do it to make my solution comply with the industries best practices.  

Reviewer: ```suggestion
But afterwards he would always take the time to explain why, and how, I should do it to make my solution comply with the industry best practices.  
```

> Being able to have someone who makes you self-reflect on the work you did, and takes his time to give you proper feedback, is the biggest asset to provide to a junior programmer.  

Reviewer: ```suggestion
Being able to have someone who makes you self-reflect on the work you did and takes his time to give you proper feedback.
That is the biggest asset to provide to a junior programmer.  
```
